README.md
# Discord RPC Client

## Examples

### Big image, small image, details and one button

![](https://i.imgur.com/6MmBURG.png)

```bash
discordrpc --client-id 942151169185316874 --details 'untypeable nickname' --button-1-text 'Discord RPC written in Rust' --button-1-url 'https://github.com/MedzikUser/discord-rpc' --large-image medzik --large-image-text medzik --small-image archlinux --small-image-text archlinux
```

### Big image, details, state and two buttons

![](https://i.imgur.com/tRbcy40.png)

```bash
discordrpc --client-id 942151169185316874 --details 'untypeable nickname' --state 'MedzikUser' --button-1-text 'Discord RPC written in Rust' --button-1-url 'https://github.com/MedzikUser/discord-rpc' --button-2-text 'GitHub' --button-2-url 'https://github.com/MedzikUser' --large-image medzik --large-image-text medzik
```

## How to install?

### Linux
Download discordrpc-linux from the [releases page](https://github.com/MedzikUser/discordrpc/releases/latest) and run

    chmod +x discordrpc-linux
    ./discordrpc-linux

### Arch Linux
Using yay ([AUR](https://aur.archlinux.org/packages/discordrpc))

    yay -S discordrpc

Alternatively you can add [this repo](https://github.com/archlinux-pkg/packages) and run

    sudo pacman -S discordrpc

### OSX
Download discordrpc-darwin from the [releases page](https://github.com/MedzikUser/discordrpc/releases/latest) and run

    chmod +x discordrpc-darwin
    ./discordrpc-darwin

### Windows
Download discordrpc-windows from the [releases page](https://github.com/MedzikUser/discordrpc/releases/latest) and run

    chmod +x discordrpc-windows
    ./discordrpc-windows

### Compile with Cargo
Make sure you have the latest version of Rust. Then you can run

    cargo install imgurs

## How do I get a client id?
1. Go to [Discord Developer Portal](https://discord.com/developers/applications)
2. Make New Application, name them freely (the name will be displayed in RPC)

![new app](https://i.imgur.com/RMUjPep.png)

3. Go to General Information

![general](https://i.imgur.com/yuQufwT.png)

3. Copy ID

![copy id](https://i.imgur.com/JDHZ6jy.png)

5. Then go to the Rich Presence tab (optional)

![rich presence](https://i.imgur.com/hIB5VEW.png)

6. Add images to be able to display them in RPC (optional)

![add image](https://i.imgur.com/vtEs7v6.png)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

<!-- next-header -->

## [Unreleased]

## [0.2.2] - 2022-05-22
- bump dependencies

## [0.2.1] - 2022-03-14
- added manpage
- change license to GPL3

## [0.2.0] - 2022-03-09
### CLI
- added completions
- update short aliases
- added option --timeout

### Other
- Added README

## [0.1.0] - 2022-02-13
### CLI
- options
    - -c, --client-id <*string*>        Discord application Client ID (required)
    - -d, --details <*string*>          Details (required)
    - -s, --state <*string*>            State (optional)
    - -N, --large-image <*string*>      Large Image name (optional)
    - -I, --large-image-text <*string*> Large Image text (optional)
    - -n, --small-image <*string*>      Small Image name (optional)
    - -i, --small-image-text <*string*> Small Image text (optional)
    - -B, --button-1-text <*string*>    Button 1 Text (optional)
    - -T, --button-1-url <*string*>     Button 1 URL address (optional)
    - -b, --button-2-text <*string*>    Button 2 Text (optional)
    - -t, --button-2-url <*string*>     Button 2 URL address (optional)
    - -e, --timer                       Enable timer (counted from the current time) (optional)

<!-- next-url -->
[Unreleased]: https://github.com/MedzikUser/discordrpc/compare/v0.2.2...HEAD
[0.2.2]: https://github.com/MedzikUser/discordrpc/commits/v0.2.2
[0.2.1]: https://github.com/MedzikUser/discordrpc/commits/v0.2.1
[0.2.0]: https://github.com/MedzikUser/discordrpc/commits/v0.2.0
[0.1.0]: https://github.com/MedzikUser/discordrpc/commits/v0.1.0


src/cli.rs
use clap::Parser;
use clap_complete::Shell;

#[derive(Debug, Parser, Clone)]
#[clap(name = "discordrpc")]
pub struct Cli {
    #[clap(
        short = 'c',
        long = "client-id",
        help = "Discord application Client ID",
        required = true,
        display_order = 1
    )]
    pub client_id: String,

    #[clap(
        short = 'd',
        long = "details",
        help = "Details",
        required = true,
        display_order = 2
    )]
    pub details: String,

    #[clap(
        short = 's',
        long = "state",
        help = "State",
        required = false,
        default_value = "none",
        display_order = 3
    )]
    pub state: String,

    #[clap(
        short = 'I',
        long = "large-image",
        help = "Large Image name",
        default_value = "none",
        required = false,
        display_order = 4
    )]
    pub large_image: String,

    #[clap(
        short = 'T',
        long = "large-image-text",
        help = "Large Image text",
        default_value = "none",
        required = false,
        display_order = 5
    )]
    pub large_text: String,

    #[clap(
        short = 'i',
        long = "small-image",
        help = "Small Image name",
        default_value = "none",
        required = false,
        display_order = 6
    )]
    pub small_image: String,

    #[clap(
        short = 't',
        long = "small-image-text",
        help = "Small Image text",
        default_value = "none",
        required = false,
        display_order = 7
    )]
    pub small_text: String,

    #[clap(
        short = 'B',
        long = "button-1-text",
        help = "Button 1 Text",
        default_value = "none",
        required = false,
        display_order = 8
    )]
    pub button_1_text: String,

    #[clap(
        short = 'U',
        long = "button-1-url",
        help = "Button 1 URL address",
        default_value = "none",
        required = false,
        display_order = 9
    )]
    pub button_1_url: String,

    #[clap(
        short = 'b',
        long = "button-2-text",
        help = "Button 2 Text",
        default_value = "none",
        required = false,
        display_order = 10
    )]
    pub button_2_text: String,

    #[clap(
        short = 'u',
        long = "button-2-url",
        help = "Button 2 URL address",
        default_value = "none",
        required = false,
        display_order = 11
    )]
    pub button_2_url: String,

    #[clap(
        short = 'e',
        long = "timer",
        help = "Enable timer (counted from the current time)",
        display_order = 12
    )]
    pub enable_time: bool,

    #[clap(
        short = 'E',
        long = "timeout",
        help = "Exit after X seconds",
        default_value = "0",
        required = false,
        display_order = 13
    )]
    pub timeout: u64,

    #[clap(
        long = "print-completions",
        value_name = "shell",
        arg_enum,
        display_order = 14
    )]
    pub print_completions: Option<Shell>,

    #[clap(long = "print-manpage", display_order = 15)]
    pub manpage: bool,
}


src/main.rs
mod cli;
mod execute;

use clap::{IntoApp, StructOpt};
use clap_complete::generate;
use colored::Colorize;
use std::{io, process::exit, thread::sleep, time::Duration};

use crate::cli::Cli;

fn main() {
    better_panic::install();

    let args = Cli::parse();

    // print completions
    if let Some(shell) = args.print_completions {
        let mut clap_app = Cli::command();
        let app_name = clap_app.get_name().to_string();

        generate(shell, &mut clap_app, app_name, &mut io::stdout());
        exit(0)
    }

    // print manpage
    if args.manpage {
        let clap_app = Cli::command();
        let man = clap_mangen::Man::new(clap_app);

        man.render(&mut io::stdout()).expect("generate manpage");
        exit(0)
    }

    // start discord rpc
    execute::run(args.clone());

    println!(
        "{} {}",
        "Connected!".green(),
        "Press Ctrl+C to exit!".magenta()
    );

    if args.timeout != 0 {
        sleep(Duration::from_secs(args.timeout));

        println!("{}", "Stopping due to timeout...".blue());

        exit(0)
    } else {
        loop {
            // empty `loop {}` wastes CPU cycles
            sleep(Duration::from_secs(9999999));
        }
    }
}


src/execute.rs
use colored::Colorize;
use discord_rich_presence::{
    activity::{self, Activity},
    DiscordIpc, DiscordIpcClient,
};
use std::{
    time::{SystemTime, UNIX_EPOCH},
    vec,
};

use crate::cli::Cli;

pub fn run(args: Cli) {
    let state = args.state;
    let details = args.details;
    let large_image = args.large_image;
    let large_text = args.large_text;
    let small_image = args.small_image;
    let small_text = args.small_text;
    let button_1_text = args.button_1_text;
    let button_1_url = args.button_1_url;
    let button_2_text = args.button_2_text;
    let button_2_url = args.button_2_url;
    let enable_timer = args.enable_time;

    let mut client = DiscordIpcClient::new(&args.client_id).expect("failed to create client");

    let activity = Activity::new();

    client
        .connect()
        .expect("failed to connect to Discord, please try again or relaunch Discord app");

    println!("{} {}", "details :".cyan(), details.yellow());

    let mut activity: Activity = activity.details(&details);

    if state != "none" {
        println!("{} {}", "state :".cyan(), state.yellow());

        activity = activity.state(&state);
    }

    let mut assets = activity::Assets::new();

    if large_image != "none" {
        println!("{} {}", "large image :".cyan(), large_image.yellow());

        assets = assets.large_image(&large_image);
    }

    if large_text != "none" {
        println!("{} {}", "large image text :".cyan(), large_text.yellow());

        assets = assets.large_text(&large_text);
    }

    if small_image != "none" {
        println!("{} {}", "small image :".cyan(), small_image.yellow());

        assets = assets.small_image(&small_image);
    }

    if small_text != "none" {
        println!("{} {}", "small image text :".cyan(), small_text.yellow());

        assets = assets.small_text(&small_text);
    }

    activity = activity.assets(assets);

    if button_1_text != "none" && button_1_url != "none" {
        println!("{} {}", "button 1 text :".cyan(), button_1_text.yellow());
        println!("{} {}", "button 1 url :".cyan(), button_1_url.yellow());

        activity = activity.buttons(vec![activity::Button::new(&button_1_text, &button_1_url)]);
    }

    if button_1_text != "none"
        && button_1_url != "none"
        && button_2_text != "none"
        && button_2_url != "none"
    {
        println!("{} {}", "button 2 text :".cyan(), button_2_text.yellow());
        println!("{} {}", "button 2 url :".cyan(), button_2_url.yellow());

        activity = activity.buttons(vec![
            activity::Button::new(&button_1_text, &button_1_url),
            activity::Button::new(&button_2_text, &button_2_url),
        ]);
    }

    if enable_timer {
        let time_unix = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;

        activity = activity.timestamps(activity::Timestamps::new().start(time_unix))
    }

    client.set_activity(activity).expect("client set activity");
}


src/cli/parse.rs
use clap::Parser;
use clap_complete::Shell;

#[derive(Debug, Parser, Clone)]
#[clap(name = "discordrpc")]
pub struct Cli {
    #[clap(
        short = 'c',
        long = "client-id",
        help = "Discord application Client ID",
        required = true,
        display_order = 1
    )]
    pub client_id: String,

    #[clap(
        short = 'd',
        long = "details",
        help = "Details",
        required = true,
        display_order = 2
    )]
    pub details: String,

    #[clap(
        short = 's',
        long = "state",
        help = "State",
        required = false,
        default_value = "none",
        display_order = 3
    )]
    pub state: String,

    #[clap(
        short = 'I',
        long = "large-image",
        help = "Large Image name",
        default_value = "none",
        required = false,
        display_order = 4
    )]
    pub large_image: String,

    #[clap(
        short = 'T',
        long = "large-image-text",
        help = "Large Image text",
        default_value = "none",
        required = false,
        display_order = 5
    )]
    pub large_text: String,

    #[clap(
        short = 'i',
        long = "small-image",
        help = "Small Image name",
        default_value = "none",
        required = false,
        display_order = 6
    )]
    pub small_image: String,

    #[clap(
        short = 't',
        long = "small-image-text",
        help = "Small Image text",
        default_value = "none",
        required = false,
        display_order = 7
    )]
    pub small_text: String,

    #[clap(
        short = 'B',
        long = "button-1-text",
        help = "Button 1 Text",
        default_value = "none",
        required = false,
        display_order = 8
    )]
    pub button_1_text: String,

    #[clap(
        short = 'U',
        long = "button-1-url",
        help = "Button 1 URL address",
        default_value = "none",
        required = false,
        display_order = 9
    )]
    pub button_1_url: String,

    #[clap(
        short = 'b',
        long = "button-2-text",
        help = "Button 2 Text",
        default_value = "none",
        required = false,
        display_order = 10
    )]
    pub button_2_text: String,

    #[clap(
        short = 'u',
        long = "button-2-url",
        help = "Button 2 URL address",
        default_value = "none",
        required = false,
        display_order = 11
    )]
    pub button_2_url: String,

    #[clap(
        short = 'e',
        long = "timer",
        help = "Enable timer (counted from the current time)",
        display_order = 12
    )]
    pub enable_time: bool,

    #[clap(
        short = 'E',
        long = "timeout",
        help = "Exit after X seconds",
        default_value="0",
        required = false,
        display_order = 13
    )]
    pub timeout: u64,

    #[clap(
        long = "print-completions",
        value_name = "shell",
        arg_enum,
        display_order = 14
    )]
    pub print_completions: Option<Shell>,

    #[clap(
        long = "print-manpage",
        display_order = 15
    )]
    pub manpage: bool,
}
